src/passport-saml/saml.ts

```diff
@@ -107,100 +107,111 @@ async function promiseWithNameID(nameid: Node): Promise<NameID> {
-      throw new Error("Invalid property: cert must not be empty");
-    }
-
-    if (!options.path) {
-      options.path = "/saml/consume";
-    }
-
-    if (!options.host) {
-      options.host = "localhost";
-    }
-
-    if (!options.issuer) {
-      options.issuer = "onelogin_saml";
-    }
-
-    if (options.identifierFormat === undefined) {
-      options.identifierFormat = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress";
-    }
-
-    if (options.authnContext === undefined) {
-      options.authnContext = "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport";
-    }
-
-    if (!Array.isArray(options.authnContext)) {
-      options.authnContext = [options.authnContext];
-    }
-
-    if (!options.acceptedClockSkewMs) {
-      // default to no skew
-      options.acceptedClockSkewMs = 0;
-    }
-
-    if (!options.validateInResponseTo) {
-      options.validateInResponseTo = false;
-    }
-
-    if (!options.requestIdExpirationPeriodMs) {
-      options.requestIdExpirationPeriodMs = 28800000; // 8 hours
-    }
-
-    if (!options.cacheProvider) {
-      options.cacheProvider = new InMemoryCacheProvider({
-        keyExpirationPeriodMs: options.requestIdExpirationPeriodMs,
-      });
-    }
-
-    if (!options.logoutUrl) {
-      // Default to Entry Point
-      options.logoutUrl = options.entryPoint || "";
-    }
-
-    // sha1, sha256, or sha512
-    if (!options.signatureAlgorithm) {
-      options.signatureAlgorithm = "sha1";
-    }
-
-    /**
-     * List of possible values:
-     * - exact : Assertion context must exactly match a context in the list
-     * - minimum:  Assertion context must be at least as strong as a context in the list
-     * - maximum:  Assertion context must be no stronger than a context in the list
-     * - better:  Assertion context must be stronger than all contexts in the list
-     */
-    if (
-      !options.RACComparison ||
-      ["exact", "minimum", "maximum", "better"].indexOf(options.RACComparison) === -1
-    ) {
-      options.RACComparison = "exact";
-    }
+    const options = {
+      ...this.ctorOptions,
+      passive: this.ctorOptions.passive ?? false,
+      disableRequestedAuthnContext: this.ctorOptions.disableRequestedAuthnContext ?? false,
+      additionalParams: this.ctorOptions.additionalParams ?? {},
+      additionalAuthorizeParams: this.ctorOptions.additionalAuthorizeParams ?? {},
+      additionalLogoutParams: this.ctorOptions.additionalLogoutParams ?? {},
+      forceAuthn: this.ctorOptions.forceAuthn ?? false,
+      skipRequestCompression: this.ctorOptions.skipRequestCompression ?? false,
+      disableRequestAcsUrl: this.ctorOptions.disableRequestAcsUrl ?? false,
+      acceptedClockSkewMs: this.ctorOptions.acceptedClockSkewMs ?? 0,
+      path: this.ctorOptions.path ?? "/saml/consume",
+      host: this.ctorOptions.host ?? "localhost",
+      issuer: this.ctorOptions.issuer ?? "onelogin_saml",
+      identifierFormat:
```

Comment:
Why not using `??` for `identifierFormat`, while you are using `??` for `authContext`?  
`identifierFormat` and `authnContext` both used `=== undefined` in the old code, so I would expect them to use the same logic in the new code.  
Also in the README the text for both states "if truthy..." which also makes me expect the same kind of logic.

Reply:
You are correct. I've reviewed the logic here and found that we can use the same `??` logic. I initially just copied the pre-existing logic. As you can see, there were a lot of places where we played fast-and-lose with what we allowed leaving a lot of room for misconfiguration, even with TypeScript.

Reply:
After reviewing some tests, it turns out that this logic is important to have written this way. If `null` is specified, then the default is over-ridden, however, if nothing is provided, or `undefined`, then we use the default. Otherwise, we use what is provided. So, for a test to pass which tests to make sure that this block of XML is removed when `null` is passed, this has to remain this way.